crates/ruff_linter/src/rules/flake8_pyi/rules/redundant_numeric_union.rs

```diff
@@ -66,57 +85,212 @@ pub(crate) fn redundant_numeric_union(checker: &mut Checker, parameters: &Parame
+    let mut remove_numeric_type = |expr: &'a Expr, parent: &'a Expr| {
+        let Some(builtin_type) = checker.semantic().resolve_builtin_symbol(expr) else {
+            // Keep type annotations that are not numeric.
+            necessary_nodes.push(expr);
+            return;
+        };
+
+        if matches!(parent, Expr::BinOp(_)) {
+            union_type = UnionKind::PEP604;
+        }
+
+        // `int` is always dropped, since `float` or `complex` must be present.
+        // `float` is only dropped if `complex`` is present.
+        if (builtin_type == "float" && !numeric_flags.contains(NumericFlags::COMPLEX))
+            || (builtin_type != "float" && builtin_type != "int")
+        {
+            necessary_nodes.push(expr);
         }
+    };
 
-        if has_int {
-            checker.diagnostics.push(Diagnostic::new(
-                RedundantNumericUnion {
-                    redundancy: Redundancy::IntComplex,
-                },
-                annotation.range(),
-            ));
+    // Traverse the union, and remember which numeric types are found.
+    traverse_union(&mut remove_numeric_type, checker.semantic(), annotation);
+
+    // Mark [`Fix`] as unsafe when comments are in range.
+    let applicability = if checker.comment_ranges().intersects(annotation.range()) {
+        Applicability::Unsafe
+    } else {
+        Applicability::Safe
+    };
+
+    // Generate the flattened fix once.
+    let fix = if let &[edit_expr] = necessary_nodes.as_slice() {
+        // Generate a [`Fix`] for a single type expression, e.g. `int`.
+        Fix::applicable_edit(
+            Edit::range_replacement(checker.generator().expr(edit_expr), annotation.range()),
+            applicability,
+        )
+    } else {
+        match union_type {
+            UnionKind::PEP604 => {
+                generate_pep604_fix(checker, necessary_nodes, annotation, applicability)
+            }
+            UnionKind::TypingUnion => {
+                generate_union_fix(checker, necessary_nodes, annotation, applicability)
+                    .ok()
+                    .unwrap()
+            }
         }
-    } else if has_float && has_int {
-        checker.diagnostics.push(Diagnostic::new(
-            RedundantNumericUnion {
-                redundancy: Redundancy::IntFloat,
-            },
-            annotation.range(),
-        ));
+    };
+
+    checker.diagnostics.push(
+        Diagnostic::new(RedundantNumericUnion { redundancy }, annotation.range()).with_fix(fix),
+    );
+}
+
+#[derive(Debug, Clone, Copy, Eq, PartialEq)]
+enum Redundancy {
+    IntFloatComplex,
+    FloatComplex,
+    IntComplex,
+    IntFloat,
+}
+
+impl Redundancy {
+    pub(super) fn from_numeric_flags(numeric_flags: NumericFlags) -> Option<Self> {
+        match numeric_flags.bits() {
+            // NumericFlags::INT | NumericFlags::FLOAT | NumericFlags::COMPLEX
+            0b0111 => Some(Self::IntFloatComplex),
```

Comment:
There must be a more elegant way of doing this, but I haven't found it yet and this works.

Reply:
I don't think there is and it's inherent to bitflags where it isn't required that all possible combinations are valid. 

I would write this as

```
        if numeric_flags == NumericFlags::INT | NumericFlags::FLOAT | NumericFlags::COMPLEX {
            Some(Self::IntFloatComplex)
        } else if numeric_flags == NumericFlags::INT | NumericFlags::FLOAT {
            Some(Self::IntComplex)
        } else if numeric_flags == NumericFlags::FLOAT | NumericFlags::COMPLEX {
            Some(Self::FloatComplex)
        } else if numeric_flags == NumericFlags::FLOAT | NumericFlags::INT {
            Some(Self::IntFloat)
        } else {
            None
        }
```

Reply:
Thanks, yes that looks better.